src/cargo/core/compiler/fingerprint.rs

```diff
@@ -1533,9 +1565,23 @@ pub fn translate_dep_info(
 
     let mut new_contents = Vec::new();
     for file in deps {
-        let absolute = rustc_cwd.join(file);
-        let path = absolute.strip_prefix(pkg_root).unwrap_or(&absolute);
-        new_contents.extend(util::path2bytes(path)?);
+        let file = PathBuf::from(file);
+        let (ty, path) = if file.is_relative() {
```

Comment:
I think this can probably be simplified to perhaps only two cases, one which is maybe package-root relative and one which is target-root-relative. The path is always calculated the same here (`rustc_cwd.join(file)`), and then if it happens to be prefixed by the target root or package root we strip it and store such, and otherwise we don't gain a huge amount from having a separate absolute path variant.

I'm also more inclined to make the path absolute as fast as possible and then consider it all relative to various locations instead of checking for it being relative first, since I think that'll help various cwd switching mechanics over time.

Reply:
I don't quite follow you here. Is the idea that we rely on `foo.join(bar)` not having any effect if `bar` is absolute?

Reply:
Right yeah (and afaik that's documented behavior). It's something I rely on basically every time I call `join` at least :)